### PR TITLE
generalize operator to support Consul

### LINKED ISSUE #32: Generalize the Operator and Support Consul
Support for Consul in istio/istio has been spotty (it only supports Consul 1.3) and [was removed from upstream](https://github.com/istio/istio/pull/25521/files) (disappearing in Istio 1.8). However, there's been quite a lot of interest in Consul as a service discovery source for Istio. The plan of record in upstream is to move to support Consul via MCP (i.e. consume it as ServiceEntries from another service).

Given asks from customers, we'll should integrate and update the Istio-Consul integration into this codebase (as well as generally stepping up maintenance here).

This will require:
- [ ] Cleaning up the [cloudmap.Store interface](https://github.com/tetratelabs/istio-cloud-map/blob/013ebc194717d6ef696e18c261337a5539caebe7/pkg/cloudmap/store.go#L11-L14)
    - Move the declaration over to the consumer, `pkg/control`
    - Do any work needed to make it play nicely with the previous Istio consul code
- [ ] Bringing the previous istio/istio Consul code over here (with git history/attribution)
    - @nacx can I get some git help from you on this?
- [ ] Update the Consul code to a modern Consul version (it uses 1.3; current version is 1.8)
    - We should investigate compatibility here - I'm not 100% sure what version of Consul it's best for us to cover - by default I assume we should target latest.
- [ ] Fit together the Store interface and the updated istio/istio Consul code

Outside of the code stuffs:
- [ ] Moving over repo name
- [ ] Updating documents
- [ ] Cut a 0.2 release before all this stuff [@ZackButcher] 
- [ ] Cut a 0.3 release after all this stuff

cc @PetrMc

### PR DESCRIPTION
**Pull request by mathetake**
resolves #32 

___

### TODOs
- [x] implement blocking request by using `WaitIndex`: done in https://github.com/tetratelabs/istio-cloud-map/pull/33/commits/ede71a229eb47faf265bdd75531bb407b3f6322d 
- [x] integrate consul with `cmd/main.go` : done in https://github.com/tetratelabs/istio-cloud-map/pull/33/commits/56e4bb39591638fd314cfccb17db22325cc0a0c5
- [x] setting up GHA: done in https://github.com/tetratelabs/istio-cloud-map/pull/33/commits/042b25da3103f3086681c7eba3a8eff7d821874e
- [x] update README to add instruction for running tests locally: done in https://github.com/tetratelabs/istio-cloud-map/pull/33/commits/e8c150c5b3277e1f8856a5491a7e7a9ff5e4a976
- [x] consul namespace support: done in https://github.com/tetratelabs/istio-cloud-map/pull/33/commits/026552fb3077988e96a5f75789f3e9f2eb3e2566

### NOTE
- I haven't borrowed codes from istio's deleted codes following the discussion with @lizan. This is because the consul's api has been changed and the implementation is not quite complicated.
- Before merging this PR, I should wait for #29 to be merged and the tag 0.2 to be released

### CODE REVIEW

**Comment by lizan** [comment]
@mathetake re: CI, GitHub Actions is our preferred tooling :)

File: pkg/consul/watcher.go [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+		if len(eps) > 0 {
+			data[name] = eps
+		}
+	}
+	w.store.Set(data)
+}
+
+// listServices lists services
+func (w *watcher) listServices() (map[string][]string, error) {
+	// TODO: support Namespace? Namespaces are available only in Consul Enterprise(+1.7.0)
```
Comment (ZackButcher) [inline pkg/consul/watcher.go:72]: Can we put it behind a flag cleanly? I suspect we'll run into Consul <1.7 in the wild fairly frequently - I'll check with some of our customers to verify.

File: pkg/consul/watcher.go [hunk 2]
```diff
@@ -0,0 +1,118 @@
+package consul
+
+import (
+	"context"
+	"log"
```
Comment (ZackButcher) [inline pkg/consul/watcher.go:5]: Use https://github.com/tetratelabs/log instead please.

File: pkg/consul/watcher.go [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+var _ provider.Watcher = &watcher{}
+
+func NewWatcher(store provider.Store, client *api.Client) provider.Watcher {
+	return &watcher{client: client, store: store, interval: time.Second * 5}
+}
+
+// Run the watcher until the context is cancelled
+func (w *watcher) Run(ctx context.Context) {
+	tick := time.NewTicker(w.interval).C
```
Comment (ZackButcher) [inline pkg/consul/watcher.go:30]: This leaks a goroutine, instead:
```go
ticker := time.NewTicker(w.interval)
defer ticker.Stop()

tick := ticker.C
// OR
select {
    case <-ticker.C: // ...
    // ...
}
```

**Review by ZackButcher — COMMENTED** [review]
Looks great! Just one or two small things in the current code.

**Comment by mathetake** [comment]
@ZackButcher thank you for review! going to reflect comments soon :smile: 

@lizan `GitHub Actions is our preferred tooling :)` got it.

File: cmd/istio-cloud-map/main.go [hunk 4]
```diff
@@ -22,12 +22,14 @@ import (
 	"strings"
 	"time"
 
+	"github.com/tetratelabs/istio-cloud-map/pkg/provider"
```
[the remark this answers is not in the document]
Comment (mathetake) [reply cmd/istio-cloud-map/main.go]: fixed in https://github.com/tetratelabs/istio-cloud-map/pull/33/commits/74e171edfcd1785da85e5be52d00e7ab35e0ad20

File: pkg/consul/watcher.go (hunk 2 above)
Comment (mathetake) [reply pkg/consul/watcher.go]: fixed in https://github.com/tetratelabs/istio-cloud-map/pull/33/commits/74e171edfcd1785da85e5be52d00e7ab35e0ad20

**Comment by mathetake** [comment]
As for CI, I will create another PR to set up GHA

File: pkg/consul/watcher.go (hunk 3 above)
Comment (mathetake) [reply pkg/consul/watcher.go]: https://github.com/tetratelabs/istio-cloud-map/pull/33/commits/8b7a0f3fa2049b86264028d2f370bac2c3a95832 fixed

File: cmd/istio-cloud-map/main.go [hunk 5]
```diff
@@ -85,28 +88,19 @@ func serve() (serve *cobra.Command) {
 			// TODO: move over to run groups, get a context there to use to handle shutdown gracefully.
 			ctx := context.Background() // common context for cancellation across all loops/routines
 
-			// TODO: see if it makes sense to push this down into the CM section after moving to run groups
-			if len(awsRegion) == 0 {
```
Comment (mathetake) [inline cmd/istio-cloud-map/main.go:88]: moveed to pkg/cloudmap

**Comment by mathetake** [comment]
it seems to start working :smile: 

```
>> k get serviceentries.networking.istio.io
NAME            HOSTS      LOCATION   RESOLUTION   AGE
consul-consul   [consul]              STATIC       11m
consul-redis    [redis]               STATIC       7m54s

>> k get serviceentries.networking.istio.io consul-redis -oyaml
apiVersion: networking.istio.io/v1beta1
kind: ServiceEntry
metadata:
  creationTimestamp: "2020-09-03T05:21:11Z"
  generation: 1
  managedFields:
  - apiVersion: networking.istio.io/v1alpha3
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:ownerReferences:
          .: {}
          k:{"uid":"87d7bcd5-bf43-41ae-8f58-24878d6d461b"}:
            .: {}
            f:apiVersion: {}
            f:controller: {}
            f:kind: {}
            f:name: {}
            f:uid: {}
      f:spec:
        .: {}
        f:addresses: {}
        f:endpoints: {}
        f:hosts: {}
        f:ports: {}
        f:resolution: {}
    manager: istio-cloud-map
    operation: Update
    time: "2020-09-03T05:21:11Z"
  name: consul-redis
  namespace: default
  ownerReferences:
  - apiVersion: cloudmap.istio.io
    controller: true
    kind: ServiceController
    name: istio-cloud-map-operator
    uid: 87d7bcd5-bf43-41ae-8f58-24878d6d461b
  resourceVersion: "10799"
  selfLink: /apis/networking.istio.io/v1beta1/namespaces/default/serviceentries/consul-redis
  uid: e839596d-8dda-4ab8-b952-c9792f8a0014
spec:
  addresses:
  - 192.168.10.10
  endpoints:
  - address: 192.168.10.10
    ports:
      tcp: 8000
  hosts:
  - redis
  ports:
  - name: tcp
    number: 8000
    protocol: TCP
  resolution: STATIC

>> k get serviceentries.networking.istio.io consul-consul -oyaml
apiVersion: networking.istio.io/v1beta1
kind: ServiceEntry
metadata:
  creationTimestamp: "2020-09-03T05:17:16Z"
  generation: 1
  managedFields:
  - apiVersion: networking.istio.io/v1alpha3
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:ownerReferences:
          .: {}
          k:{"uid":"87d7bcd5-bf43-41ae-8f58-24878d6d461b"}:
            .: {}
            f:apiVersion: {}
            f:controller: {}
            f:kind: {}
            f:name: {}
            f:uid: {}
      f:spec:
        .: {}
        f:addresses: {}
        f:endpoints: {}
        f:hosts: {}
        f:ports: {}
        f:resolution: {}
    manager: istio-cloud-map
    operation: Update
    time: "2020-09-03T05:17:16Z"
  name: consul-consul
  namespace: default
  ownerReferences:
  - apiVersion: cloudmap.istio.io
    controller: true
    kind: ServiceController
    name: istio-cloud-map-operator
    uid: 87d7bcd5-bf43-41ae-8f58-24878d6d461b
  resourceVersion: "10064"
  selfLink: /apis/networking.istio.io/v1beta1/namespaces/default/serviceentries/consul-consul
  uid: 9d8b92dd-b2a2-450c-8eca-0539db54d690
spec:
  addresses:
  - 127.0.0.1
  endpoints:
  - address: 127.0.0.1
    ports:
      tcp: 8300
  hosts:
  - consul
  ports:
  - name: tcp
    number: 8300
    protocol: TCP
  resolution: STATIC

```

These are the results of the following consul catalog state: 

```
>> curl http://127.0.0.1:8500/v1/catalog/services
{
    "consul": [],
    "redis": [
        "primary",
        "v1"
    ]
}

>> curl http://127.0.0.1:8500/v1/catalog/service/redis
[
    {
        "ID": "40e4a748-2192-161a-0510-9bf59fe950b5",
        "Node": "foobar",
        "Address": "192.168.10.10",
        "Datacenter": "dc1",
        "TaggedAddresses": {
            "lan": "192.168.10.10",
            "wan": "10.0.10.10"
        },
        "NodeMeta": {
            "somekey": "somevalue"
        },
        "ServiceKind": "",
        "ServiceID": "redis1",
        "ServiceName": "redis",
        "ServiceTags": [
            "primary",
            "v1"
        ],
        "ServiceAddress": "127.0.0.1",
        "ServiceTaggedAddresses": {
            "lan": {
                "Address": "127.0.0.1",
                "Port": 8000
            },
            "wan": {
                "Address": "198.18.0.1",
                "Port": 80
            }
        },
        "ServiceWeights": {
            "Passing": 1,
            "Warning": 1
        },
        "ServiceMeta": {
            "redis_version": "4.0"
        },
        "ServicePort": 8000,
        "ServiceEnableTagOverride": false,
        "ServiceProxy": {
            "MeshGateway": {},
            "Expose": {}
        },
        "ServiceConnect": {},
        "CreateIndex": 130,
        "ModifyIndex": 130
    }
]

>> curl http://127.0.0.1:8500/v1/catalog/service/consul
[
    {
        "ID": "df3cac03-4b06-6b41-ede9-ad954ed12d5c",
        "Node": "53b96b8a5479",
        "Address": "127.0.0.1",
        "Datacenter": "dc1",
        "TaggedAddresses": {
            "lan": "127.0.0.1",
            "lan_ipv4": "127.0.0.1",
            "wan": "127.0.0.1",
            "wan_ipv4": "127.0.0.1"
        },
        "NodeMeta": {
            "consul-network-segment": ""
        },
        "ServiceKind": "",
        "ServiceID": "consul",
        "ServiceName": "consul",
        "ServiceTags": [],
        "ServiceAddress": "",
        "ServiceWeights": {
            "Passing": 1,
            "Warning": 1
        },
        "ServiceMeta": {
            "non_voter": "false",
            "raft_version": "3",
            "serf_protocol_current": "2",
            "serf_protocol_max": "5",
            "serf_protocol_min": "1",
            "version": "1.8.3"
        },
        "ServicePort": 8300,
        "ServiceEnableTagOverride": false,
        "ServiceProxy": {
            "MeshGateway": {},
            "Expose": {}
        },
        "ServiceConnect": {},
        "CreateIndex": 11,
        "ModifyIndex": 11
    }
]
```

File: pkg/consul/watcher.go (hunk 1 above)
Comment (mathetake) [reply pkg/consul/watcher.go]: https://github.com/tetratelabs/istio-cloud-map/pull/33/commits/026552fb3077988e96a5f75789f3e9f2eb3e2566 done :smile:

**Comment by mathetake** [comment]
@ZackButcher, (cc @lizan)  I've just made the PR ready, so PTAK when you have time

File: pkg/consul/watcher.go [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+	url, err := url.Parse(endpoint)
+	if err != nil {
+		return nil, errors.Wrap(err, "error parsing endpoint")
+	}
+
+	// TODO: allow users to specify TOKEN
+	// TODO: Since namespace feature is only available in Enterprise (+1.7.0), we haven't tested yet
+	config.Scheme = url.Scheme
+	config.Address = url.Host
+	config.WaitTime = 5 * time.Second
```
Comment (lizan) [inline pkg/consul/watcher.go:44]: Will this works for all case? add a TODO to make this (and tickInterval below) configurable?

File: pkg/consul/watcher_test.go [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+		services map[string][]*api.CatalogRegistration
+	}{
+		{
+			name: "single",
+			services: map[string][]*api.CatalogRegistration{
+				"service1": {
+					{
+						ID:      "aaaaaaaa-aaaa-aaaa-aaaa-aaaaaaaaaaaa",
+						Node:    "node1",
+						Address: "1.1.1.1",
```
Comment (lizan) [inline pkg/consul/watcher_test.go:63]: nit: use IP addresses from document/example range (192.0.2.0/24) or private IP range? (throughout)

Comment (mathetake) [reply pkg/consul/watcher_test.go]: https://github.com/tetratelabs/istio-cloud-map/pull/33/commits/7097dcf75beefeb1a087e989778e6a51b1a03a7f done!

File: pkg/consul/watcher.go (hunk 6 above)
Comment (mathetake) [reply pkg/consul/watcher.go]: done! https://github.com/tetratelabs/istio-cloud-map/pull/33/commits/7097dcf75beefeb1a087e989778e6a51b1a03a7f

**Review by lizan — APPROVED** [review]

**Pull request merged**